src/fairseq2/nn/lora.py

```diff
@@ -0,0 +1,373 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import math
+import re
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Any, Dict, List, Literal, Optional, final
+
+import torch
+import torch.nn as nn
+from torch import Tensor
+from torch.nn import Dropout
+from torch.nn.functional import embedding, linear
+from torch.nn.parameter import Parameter
+
+from fairseq2.nn import Embedding, Linear, Projection, StandardEmbedding
+from fairseq2.typing import DataType, Device
+
+
+@dataclass
+class LoRAConfig:
+    r: int
+    alpha: float
+    dropout_p: float
+    keys: List[str]
+
+
+class LoRALayer(ABC):
+    def __init__(self, config: LoRAConfig):
+        self.r = config.r
+        self.alpha = config.alpha
+        self.scaling = self.alpha / self.r
+        self.dropout_p = config.dropout_p
+
+    @property
+    @abstractmethod
+    def wrapped_module(self) -> nn.Module:
+        ...
+
+    @abstractmethod
+    def merge(self) -> None:
+        ...
+
+    @abstractmethod
+    def unmerge(self) -> None:
+        ...
+
+
+@final
+class LoRAEmbedding(Embedding, LoRALayer):
+    weight: Parameter
+    lora_A: Parameter
+    lora_B: Parameter
+    merged: bool
+
+    def __init__(
+        self,
+        wrapped: Embedding,
+        config: LoRAConfig,
+        device: Optional[Device] = None,
+        dtype: Optional[DataType] = None,
+    ) -> None:
+        """
+        :param wrapped:
+            The Embedding module to be wrapped.
+        :param config:
+            The LoRA config.
+        """
+        Embedding.__init__(
+            self,
+            wrapped.num_embeddings,
+            wrapped.embedding_dim,
+            wrapped.pad_idx,
+        )
+        LoRALayer.__init__(self, config)
+
+        self.weight = Parameter(
+            wrapped.weight, requires_grad=False  # type: ignore[arg-type]
+        )
+
+        self.lora_A = Parameter(
+            torch.empty((self.r, self.num_embeddings), device=device, dtype=dtype)
+        )
+
+        self.lora_B = Parameter(
+            torch.empty((self.embedding_dim, self.r), device=device, dtype=dtype)
+        )
+
+        self.merged = False
+
+        self.reset_lora_parameters()
+
+    def forward(self, x: Tensor) -> Tensor:
+        if self.merged:
+            return embedding(x, self.weight, self.pad_idx)
+        else:
+            return embedding(
+                x,
+                self.weight + (self.lora_B @ self.lora_A).T * self.scaling,
+                self.pad_idx,
+            )
+
+    def reset_lora_parameters(self) -> None:
+        """Reset the parameters and buffers of the module."""
+        nn.init.zeros_(self.lora_A)
+        nn.init.normal_(self.lora_B)
+
+    @property
+    def wrapped_module(self) -> nn.Module:
+        return self
```

Comment:
You should ideally store the `wrapped` argument passed to the constructor as a state variable and return it here.

```python
class LoRAEmbedding(Embedding, LoRALayer):
  def __init__(self, wrapped: Embedding, ...) -> None:
    self.wrapped = wrapped

  @property
   def wrapped_module(self) -> Embedding:
     return self.wrapped
```

Reply:
I think this also addresses your question in the PR summary about wrap/unwrap discrepancy about the model types. Ideally we should always preserve the original wrapped module within the LoRA module.